amandaguan-ag/todo
# #1 basic task management

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|  ✓    | :sparkles: New feature     

## Acceptance Criteria

- [x] Implement a UI component for displaying tasks.
- [x] Create a form component for adding new tasks (input field and submit button).
- [x] Set up state management for handling the addition, completion, and deletion of tasks.
- [x] Design and implement a database schema for tasks (fields: id, title, completed).
- [x] Write backend API endpoints for creating, updating, and deleting tasks.
- [x] Implement frontend logic for interacting with these API endpoints.

## Related Issue

Closes #1 

## Description

This PR introduces significant frontend enhancements and connects the frontend to the backend NestJS server for data persistence.

### Display Tasks from State

I have implemented a `TaskList` component that uses the local state to display tasks. It updates the UI in real time as tasks are added or marked complete.

### Write Tasks to the Backend

With the new `AddTask` component, users can now add tasks that are sent to the backend server and persisted in the database. Additionally, tasks can now be marked as completed or deleted, with these changes immediately updating the state and reflecting in the UI.

### Testing Steps / QA Criteria

To prepare for testing, please ensure you've followed the [setup steps](https://github.com/amandaguan-ag/todo#setup) outlined in the project's README, including running any necessary database migrations to update your local database schema.

1. Checkout to this branch with `git checkout 1-basic-task-management`.
2. Make sure you are in the backend directory and run the database migrations to update your schema by executing `npm run typeorm migration:run`.
3. Start the backend service by running `npm run start` in the backend directory.
4. In a separate terminal, navigate to the frontend directory and run `npm start` to launch the frontend.
5. Navigate to [http://localhost:3000](http://localhost:3000) to access the Todo List Application.
6. Add a new task using the form and verify it appears in the task list.
7. Mark a task as complete and ensure the UI reflects this change.
8. Delete a task and confirm it disappears from the list.
9. Verify that all changes are synchronized with the backend by checking the database.

## TODOs

- Automated tests for the new features are yet to be added. Integration tests using tools like Jest for the backend and React Testing Library for the frontend are planned.
- Error handling and form validation need to be reviewed and potentially enhanced to handle more edge cases.
- The UI can be further improved for better user experience, particularly on mobile devices.
